File: Examples/iOS/SongLibraryPlayer/SongLibraryPlayer/SongsViewController.m
//
//  SongsViewController.m
//  SongLibraryPlayer
//
//  Created by Aurelius Prochazka on 12/19/13.
//  Copyright (c) 2013 Hear For Yourself. All rights reserved.
//

#import "SongsViewController.h"
#import "SongViewController.h"
#import <AVFoundation/AVFoundation.h>
#import <MediaPlayer/MediaPlayer.h>

@interface SongsViewController () <UITableViewDelegate> {
}

@property (strong, nonatomic) NSMutableArray *songsList;

@end

@implementation SongsViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    MPMediaPropertyPredicate *artistPredicate =
    [MPMediaPropertyPredicate predicateWithValue:self.albumName
                                     forProperty:MPMediaItemPropertyAlbumTitle
                                  comparisonType:MPMediaPredicateComparisonContains];
    
    MPMediaQuery *songsQuery = [MPMediaQuery songsQuery];
    [songsQuery addFilterPredicate:artistPredicate];
    
    NSArray *itemsFromGenericQuery = [songsQuery items];
    self.songsList = [NSMutableArray arrayWithArray:itemsFromGenericQuery];
    [self.tableView reloadData];
    

}



#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView
 numberOfRowsInSection:(NSInteger)section
{
    return self.songsList.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView
         cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"SongCell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    MPMediaItem *song = [self.songsList objectAtIndex:indexPath.row];
    NSString *songTitle = [song valueForProperty: MPMediaItemPropertyTitle];
    int minutes = [[song valueForProperty:MPMediaItemPropertyPlaybackDuration] floatValue] / 60;
    int seconds = [[song valueForProperty:MPMediaItemPropertyPlaybackDuration] intValue] % 60;
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%d:%02d", minutes, seconds];
    cell.textLabel.text = songTitle;
    return cell;
}

#pragma mark - Navigation

// In a story board-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    NSIndexPath *path = [self.tableView indexPathForSelectedRow];
    SongViewController *vc = [segue destinationViewController];
    vc.song  = [self.songsList objectAtIndex:path.row];
    vc.title = [vc.song valueForProperty:MPMediaItemPropertyTitle];
}

@end

File: Examples/iOS/SongProcessor/SongProcessor/View Controllers/iTunes Library Access/SongsViewController.swift
//
//  SongsViewController.swift
//  SongProcessor
//
//  Created by Aurelius Prochazka on 6/22/16.
//  Copyright © 2016 AudioKit. All rights reserved.
//

import AVFoundation
import MediaPlayer
import UIKit

class SongsViewController: UITableViewController {

    var albumName: String!
    var artistName: String!
    var songsList = [MPMediaItem]()

    override func viewDidLoad() {
        super.viewDidLoad()

        let albumPredicate = MPMediaPropertyPredicate(value: albumName,
                                                      forProperty: MPMediaItemPropertyAlbumTitle,
                                                      comparisonType: .contains)

        let artistPredicate = MPMediaPropertyPredicate(value: artistName,
                                                       forProperty: MPMediaItemPropertyArtist,
                                                       comparisonType: .contains)

        let songsQuery = MPMediaQuery.songs()
        songsQuery.addFilterPredicate(albumPredicate)
        songsQuery.addFilterPredicate(artistPredicate)
        songsQuery.addFilterPredicate(MPMediaPropertyPredicate(
            value: NSNumber(value: false as Bool),
            forProperty: MPMediaItemPropertyIsCloudItem))

        if songsQuery.items != nil {
            songsList = songsQuery.items!
            tableView.reloadData()
        }
    }

    // MARK: - Table view data source

    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }

    override func tableView(_ tableView: UITableView,
                            numberOfRowsInSection section: Int) -> Int {
        return songsList.count
    }
    override func tableView(_ tableView: UITableView,
                            cellForRowAt indexPath: IndexPath) -> UITableViewCell {

        let cellIdentifier = "SongCell"
        let cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier) ?? UITableViewCell(style: .default,
                                                                                                    reuseIdentifier: cellIdentifier)

        let song: MPMediaItem = songsList[(indexPath as NSIndexPath).row]
        let songTitle = song.value(forProperty: MPMediaItemPropertyTitle) as! String

        let minutes = (song.value(forProperty: MPMediaItemPropertyPlaybackDuration)! as AnyObject).floatValue / 60
        let seconds = ((song.value(forProperty: MPMediaItemPropertyPlaybackDuration)! as AnyObject).floatValue).truncatingRemainder(dividingBy: 60)

        cell.textLabel?.text = songTitle
        cell.detailTextLabel?.text = String(format: "%.0f:%02.0f", minutes, seconds)

        return cell
    }

    // MARK: - Navigation

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {

        if segue.identifier == "SongSegue" {
            if let indexPath = tableView.indexPathForSelectedRow {
                let songVC = segue.destination as! SongViewController
                songVC.song = songsList[(indexPath as NSIndexPath).row]
                songVC.title = songVC.song!.value(forProperty: MPMediaItemPropertyTitle) as? String
            }
        }

    }
}
